lib/sshutils/networking/networking.go

```diff
@@ -0,0 +1,335 @@
+	if err := p.cmd.Start(); err != nil {
+		p.conn.Close()
+		return trace.Wrap(err)
+	}
+
+	// The child process writes errors to the parent connection for logging purposes.
+	go func() {
+		for {
+			buf := make([]byte, RequestBufferSize)
+			n, err := p.conn.Read(buf)
+			if err != nil {
+				if utils.IsOKNetworkError(err) {
+					return
+				}
+				slog.WarnContext(ctx, "Failed to read error from networking process.", "error", err)
+				return
+			}
+
+			if n > 0 {
+				slog.WarnContext(ctx, "Received unexpected error from networking process.", "error", string(buf[:n]))
+			}
+		}
+	}()
+
+	go func() {
+		defer close(p.done)
+		defer p.conn.Close()
+		// Ensure unexpected cmd failures get logged.
+		if err := p.cmd.Wait(); err != nil && !p.killed.Load() {
+			slog.WarnContext(ctx, "Networking process exited early with unexpected error.", "error", err)
+		}
+	}()
+
+	return nil
+}
+
+// Close stops the process and frees up its related resources.
+func (p *Process) Close() error {
+	p.conn.Close()
+	select {
+	case <-p.done:
+	case <-time.After(time.Second * 3):
+		slog.WarnContext(context.Background(), "Forcibly terminating networking subprocess.")
+		p.killed.Store(true)
+		p.cmd.Process.Kill()
+	}
+	return nil
+}
+
+// Dial requests a network connection from the networking subprocess.
+func (p *Process) Dial(ctx context.Context, network string, addr string) (net.Conn, error) {
+	connFD, err := p.sendRequest(ctx, Request{
+		Operation: NetworkingOperationDial,
+		Network:   network,
+		Address:   addr,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer connFD.Close()
+
+	conn, err := net.FileConn(connFD)
+	return conn, trace.Wrap(err)
+}
+
+// Listen requests a local listener from the networking subprocess.
+func (p *Process) Listen(ctx context.Context, network string, addr string) (net.Listener, error) {
+	listenerFD, err := p.sendRequest(ctx, Request{
+		Operation: NetworkingOperationListen,
+		Network:   network,
+		Address:   addr,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer listenerFD.Close()
+
+	listener, err := net.FileListener(listenerFD)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if unixListener, ok := listener.(*net.UnixListener); ok {
+		unixListener.SetUnlinkOnClose(true)
+	}
+
+	return listener, nil
+}
+
+// ListenAgent requests a local ssh-agent listener from the networking subprocess.
+func (p *Process) ListenAgent(ctx context.Context) (net.Listener, error) {
+	listenerFD, err := p.sendRequest(ctx, Request{
+		Operation: NetworkingOperationListenAgent,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer listenerFD.Close()
+
+	listener, err := net.FileListener(listenerFD)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	unixListener, ok := listener.(*net.UnixListener)
+	if !ok {
+		return nil, trace.BadParameter("expected *net.UnixListener but got %T", listener)
+	}
+
+	unixListener.SetUnlinkOnClose(true)
```

Comment:
We have to leave the unlink to the child process' graceful termination, right?
```suggestion
```

Reply:
If the parent process can unlink the listener it should and can - the child will still try to cleanup the socket files if the unlink fails due to pam_namespace

Reply:
The parent process (running as root, most likely) should not be unlinking paths that might belong to other sockets tho, and when fs namespaces are in the mix, there's no telling what the path local to the namespace actually is from the parent's point of view - and there's no telling if something else is there instead!

Reply:
Good point...